File: Binary_Tree/515.Find_Largest_Value_in_Each_Tree_Row/Python/515.Find_Largest_Value_in_Each_Tree_Row.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def largestValues(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[int]
        """
        #============================================#
        # Breath-first search based traversal method #
        #============================================#
        if (not root): #Issue/Boundary-case handling
            return []

        ############
        #Initialize
        ##### Record queue #####
        record_queue = [root]

        ##### Result array #####
        res_arry = []


        ##########################################
        #Breath-first search based loop traversal
        while (record_queue):
            len_queue = len(record_queue) #Length of queue

            record_max_val = float("-inf") #Record maximun value 

            for queue_idx in range(len_queue):
                record_root = record_queue.pop(0) #Record root

                record_max_val = (record_root.val if (record_root.val > record_max_val) else record_max_val) #Keep updating/overwriting

                ##### Check if the next root left-half node existed or not #####
                if (record_root.left):
                    record_queue.append(record_root.left) #Keep updating/recording
                else:
                    pass

                ##### Check if the next root right-half node existed or not #####
                if (record_root.right):
                    record_queue.append(record_root.right) #Keep updating/recording
                else:
                    pass

            res_arry.append(record_max_val) #Keep updating/recording

        return res_arry

File: Binary_Tree/515.Find_Largest_Value_in_Each_Tree_Row/C++/515.Find_Largest_Value_in_Each_Tree_Row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode * root) 
    {
        //=================================//
        // Input type:                     //
        //   - TreeNode * (root tree-node) //
        // Return type:                    //
        //   - vector<int> (result vector) //
        //=================================//
        if (root == nullptr) { return {}; } //Issue, Boundary-case handling

        /*Initialize*/
        ///// Record queue /////
        queue<TreeNode * > record_queue;

        ///// Result vector /////
        vector<int> res_vect;


        /*Breath-first search based loop traversal*/
        record_queue.emplace(root); //Update, Record

        while (! record_queue.empty()) //Cycled-times
        {
            int record_max_val = INT_MIN; //Record maximun value

            for (int queue_idx = record_queue.size(); (queue_idx > 0); (queue_idx--)) //Queue-size
            {
                TreeNode * record_root = record_queue.front(); record_queue.pop(); //Keep updating, popped

                record_max_val = ((record_root->val) > record_max_val) ? (record_root->val) : record_max_val; //Keep updating, overwriting

                ///// Check if the next root left-half node existed or not /////
                if ((record_root->left) != nullptr) { record_queue.emplace((record_root->left)); } //Keep updating, recording
                else { ; }

                ///// Check if the next root right-half node existed or not /////
                if ((record_root->right) != nullptr) { record_queue.emplace((record_root->right)); } //Keep updating, recording
                else { ; }

            } //Queue-size

            res_vect.emplace_back(record_max_val); //Keep updating, recording
             
        } //Cycled-times

        return res_vect;
    }
};
